### PR TITLE
fix: #678 refresh the engine configuration if settings change

### PR DESCRIPTION
**Pull request by mathPi**
To fix issue #678, we need to refresh the engine parameters as the new settings can impact those parameters.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new method for refreshing window preferences, enhancing user interface responsiveness.
	- Added a validation method that improves form handling and updates the main GUI upon user interaction.
	- Implemented a method to close dialogs while ensuring UI refresh, providing immediate feedback on preference changes.

- **Improvements**
	- Enhanced modularity in the GUI component initialization process for better maintainability and consistency of displayed information.

### CODE REVIEW

File: src/main/python/main/ayab/preferences.py
Comment (jonathanperret) [inline src/main/python/main/ayab/preferences.py:229]: The way the Preferences dialog is currently setup, the `OK` button is a lie. Each setting is updated as soon as you change its value, event though nothing happens immediately in the main window because the main event loop is blocked by the popup. You can verify this by closing the Preferences dialog without clicking `OK`, i.e. press Escape or click the window close button. Demonstration:

https://github.com/user-attachments/assets/970f7fde-77dd-4134-9bc1-0e202594be91

Therefore, intercepting the click on the `OK` button would make the UI refresh only for users that click it, causing weird behaviors.

The Preferences dialog should probably be reworked to actually commit settings only when `OK` is clicked, but that's a bigger piece of work.

Comment (mathPi) [reply src/main/python/main/ayab/preferences.py:229]: Ohhh yes, I missed that :( I'll try to find an other solution.

Comment (mathPi) [reply src/main/python/main/ayab/preferences.py]: @jonathanperret the done method seems to be triggered in all cases that allow to close the window, (ESC, close button, OK button). Do you see something else ?

**Comment by jonathanperret** [comment]
There are already two different places (that I could find!) where the UI is refreshed following a preferences update:

1. In `preferences.py`, right after the (blocking) call that opens the Preferences dialog, an `image_resizer` signal is (sometimes) emitted: https://github.com/AllYarnsAreBeautiful/ayab-desktop/blob/a03774e6a62410f3c4e57394d949b8b803860ac4/src/main/python/main/ayab/preferences.py#L194-L198
2. In `ayab.py`, there's a call to `scene.refresh()` after the (blocking) call to the above method: https://github.com/AllYarnsAreBeautiful/ayab-desktop/blob/a03774e6a62410f3c4e57394d949b8b803860ac4/src/main/python/main/ayab/ayab.py#L129-L131

I think it would be good not to add a third place to that list…

**Comment by mathPi** [comment]
> There are already two different places (that I could find!) where the UI is refreshed following a preferences update:
> 
> 1. In `preferences.py`, right after the (blocking) call that opens the Preferences dialog, an `image_resizer` signal is (sometimes) emitted: https://github.com/AllYarnsAreBeautiful/ayab-desktop/blob/a03774e6a62410f3c4e57394d949b8b803860ac4/src/main/python/main/ayab/preferences.py#L194-L198
> 2. In `ayab.py`, there's a call to `scene.refresh()` after the (blocking) call to the above method: https://github.com/AllYarnsAreBeautiful/ayab-desktop/blob/a03774e6a62410f3c4e57394d949b8b803860ac4/src/main/python/main/ayab/ayab.py#L129-L131
> 
> I think it would be good not to add a third place to that list…

Ohh yes, the second one is perfect, I update the PR. Is it better now ?

File: src/main/python/main/ayab/engine/engine.py
Comment (jonathanperret) [inline src/main/python/main/ayab/engine/engine.py:73]: This causes `self.config.refresh()` to be called twice in `__init__`, maybe this could be improved?

**Comment by jonathanperret** [comment]
> Ohh yes, the second one is perfect, I update the PR. Is it better now ?

Yes, that's definitely better (less code, yay). Although the Qt way would be to use a signal so that things are less coupled (e.g. `Engine` could listen to a `preferences_changed` signal), but that's not how the code is structured now unfortunately, so that's another change that will have to wait.

There is a change in behavior introduced by this though. The `Engine.refresh()` method is slightly misnamed as it actually resets (most) options to their default values. With the previous code, the `Infinite repeat` checkbox for example was not reset to the default when closing the Preferences dialog. This is probably not a big deal, just maybe a bit surprising.
In any case, the same options are also reset when opening a new image, which I find annoying by the way — sometimes you just want to reload the image that you've modified in an editor — so this new behavior wouldn't be so much out of line with the rest.

I have one last nit to pick with the name of that new method: I'm not sure `refresh_settings_change` makes sense as a phrase, but maybe I'll let a native English speaker give their opinion on that one. I'd have gone with `reload_settings` maybe?

Finally, thank you for taking my feedback into account but I don't have permission to approve PRs in this repo, you'll have to wait for a maintainer to drop by.

File: src/main/python/main/ayab/preferences.py
Comment (jonathanperret) [reply src/main/python/main/ayab/preferences.py]: For anyone reading this later, [my later comment](https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/700#issuecomment-2308759538) and @mathPi 's following changes obviate this discussion.

**Comment by dl1com** [comment]
> I have one last nit to pick with the name of that new method: I'm not sure `refresh_settings_change` makes sense as a phrase, but maybe I'll let a native English speaker give their opinion on that one. I'd have gone with `reload_settings` maybe?

Waiting for clarification on this, then merge

**Review by dl1com — APPROVED** [review]

**Pull request merged**
